[feature] render new options for rebalance server

himanish-star opened the pull request:
### Description

We now have a new set of options both basic and advanced that need to be shown in the UI with
- default values
- description
- warnings
- suggestions

### Fix

This PR builds the new UI which modifies the `step 1` of selecting options

https://github.com/user-attachments/assets/774dcce4-a556-4556-9d16-0a6410b3ffd5

himanish-star commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerOptions.ts:
```diff
@@ -0,0 +1,145 @@
+    allowedValues?: string[];
+}
+
+export const rebalanceServerOptions: RebalanceServerOption[] = [
+    {
+        "name": "dryRun",
+        "defaultValue": false,
+        "label": "Dry Run",
+        "type": "BOOL",
+        "description": "If enabled, rebalance will not run but expected changes that will occur will be returned",
+        "isAdvancedConfig": false,
+        "isStatsGatheringConfig": true,
+        "hasBreakingChange": false
+    },
+    {
+        "name": "preChecks",
+        "defaultValue": false,
+        "type": "BOOL",
+        "label": "Pre-Checks",
+        "description": "If enabled, will perform some pre-checks to ensure rebalance is safe, must enable dryRun to enable this",
```
@somandal you can now configure what to show in the `UI` directly from this config file

jayeshchoudhary replied:
that's the way to go!

jayeshchoudhary replied:
can we make type: "BOOL" ... as enum ?

somandal commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerOptions.ts:
```diff
@@ -0,0 +1,145 @@
+    defaultValue: string | boolean | number;
+    isAdvancedConfig: boolean;
+    isStatsGatheringConfig: boolean;
+    hasBreakingChange: boolean;
```
can we rename this?
perhaps something like: doubleCheckConfigValue

somandal replied:
Also, can we show a tool-tip for all the options marked as `hasBreakingChange` to explain why? I can share the "why" text in the doc we have if this is possible

subCode321 replied:
Yes, please share what to show in the tooltip

himanish-star replied:
Yes, please share what to show in the tooltip

somandal replied:
I've updated the reference doc with some details on the tooltip, please take a look when you get a chance. thanks!

himanish-star replied:
<img width="926" alt="image" src="https://github.com/user-attachments/assets/35664a9f-4424-4764-b9e7-ffa0deb47c89" />

somandal replied:
thanks for adding the tooltip!!

nit: any chance we can rename the field too: `hasBreakingChange` -> `addWarningButton` or something like that?
These aren't really breaking changes

jayeshchoudhary commented:
> <img alt="image" width="868" src="https://private-user-images.githubusercontent.com/30361728/422021231-f35b6b18-3adf-4d1b-9004-9a1dfb37381b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4NzUzODMsIm5iZiI6MTc0MTg3NTA4MywicGF0aCI6Ii8zMDM2MTcyOC80MjIwMjEyMzEtZjM1YjZiMTgtM2FkZi00ZDFiLTkwMDQtOWExZGZiMzczODFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDE0MTEyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlMDE2ZDc5MjI0YzQ3YThiNjhmYWUyN2M0Mzk3YmNkYjA1ZmQxNDZmNDZhN2ExOGY4NGI2MmNkNGE4ODYwMDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bcJhbafauA6Ls-r5lGJWmCFDZ6JWF9vXFTVCSfE1M4o">

Couple of UX feedback
1. let's add a divider after each section.
2. section heading and input label text looks almost same. can we make input label a bit more  small / fontWeight light?
3. Preparation checks when added in alert info seems to be invisible (due to same blue color). recommend removing checks outside of alert.
4. let's use TextField variant="outlined"
5. can we add a tooltip on hover with the reason for breaking changes? I think Sonam added reason for each breaking change input in the document.

jayeshchoudhary commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerConfigurationOptions/RebalanceServerConfigurationOptionInteger.tsx:
```diff
@@ -0,0 +1,33 @@
+        <Box display='flex' flexDirection='column'>
+            <FormControl fullWidth={true}>
+                <RebalanceServerConfigurationOptionLabel option={option} />
+                <Input
```
use <TextField type="number" />

jayeshchoudhary reviewed (commented):
lgtm overall, added minot comments
also, can you add ss for advance config?

himanish-star commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerOptions.ts:
```diff
@@ -0,0 +1,150 @@
+    isStatsGatheringConfig: boolean;
+    hasBreakingChange: boolean;
+    allowedValues?: string[];
+    toolTip?: string;
```
Added this @somandal

himanish-star commented:
@jayeshchoudhary addressed all comments:

- [x] let's add a divider after each section.
- [x] section heading and input label text looks almost same. can we make input label a bit more small / fontWeight light?
Preparation checks when added in alert info seems to be invisible (due to same blue color). recommend removing checks outside of alert.
   <img width="300" alt="image" src="https://github.com/user-attachments/assets/574d6c5a-d20a-405d-9b5d-432fbc8a8679" />


- [x] let's use TextField variant="outlined"
- [x] can we add a tooltip on hover with the reason for breaking changes? I think Sonam added reason for each breaking change input in the document.
- [x] advanced options preview
   <img width="300" alt="image" src="https://github.com/user-attachments/assets/c1cc1b59-8fc4-415a-b278-077974b50dc1" />

somandal commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerOptions.ts:
```diff
@@ -0,0 +1,150 @@
+        "isAdvancedConfig": false,
+        "isStatsGatheringConfig": false,
+        "hasBreakingChange": true,
+        "toolTip": "Enabling can cause downtime even if replication > 1"
```
nit: let's reword: "Enabling can cause downtime if replication > 1"

himanish-star replied:
Sure

jayeshchoudhary reviewed (approved):
lgtm :shipit:

somandal commented:
let's also update the latest video in the PR description? thanks!

yashmayya commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerConfigurationOptions/RebalanceServerConfigurationOptionInteger.tsx:
```diff
@@ -0,0 +1,39 @@
+import {Box, FormControl, TextField, Typography} from "@material-ui/core";
```
The linter is failing due to the missing headers in all these new files. You should be able to fix this by running `mvn license:format` or else copying the license headers from other files manually.

himanish-star replied:
updated

himanish-star replied:
it's passing now
